providers/ldap/ldap_provider.go

```diff
@@ -0,0 +1,620 @@
+package ldap
+
+import (
+	"bytes"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/rancher/go-rancher/v2"
+	"github.com/rancher/rancher-auth-service/model"
+	"gopkg.in/ldap.v2"
+)
+
+type SearchConfig struct {
+	Server               string
+	Port                 int64
+	BindDN               string
+	BindPassword         string
+	UserSearchAttributes []string
+	GroupSeachAttributes []string
+}
+
+var searchConfig *SearchConfig
+var nilIdentity = client.Identity{Resource: client.Resource{
+	Type: "identity",
+}}
+var nilToken = model.Token{Resource: client.Resource{
+	Type: "token",
+}}
+
+func (l *LClient) InitializeSearchConfig() *SearchConfig {
+	c := l.ConstantsConfig
+	return &SearchConfig{
+		Server:       l.Config.Server,
+		Port:         l.Config.Port,
+		BindDN:       l.Config.ServiceAccountUsername,
+		BindPassword: l.Config.ServiceAccountPassword,
+		UserSearchAttributes: []string{c.MemberOfAttribute,
+			c.ObjectClassAttribute,
+			l.Config.UserObjectClass,
+			l.Config.UserLoginField,
+			l.Config.UserNameField,
+			l.Config.UserSearchField,
+			l.Config.UserEnabledAttribute},
+		GroupSeachAttributes: []string{c.MemberOfAttribute,
+			c.ObjectClassAttribute,
+			l.Config.GroupObjectClass,
+			l.Config.UserLoginField,
+			l.Config.GroupNameField,
+			l.Config.GroupSearchField},
+	}
+}
+
+func (l *LClient) newConn() (*ldap.Conn, error) {
+	log.Debug("Now creating Ldap connection")
+	lConn, err := ldap.Dial("tcp", fmt.Sprintf("%s:%d", searchConfig.Server, searchConfig.Port))
+	if err != nil {
+		return nil, fmt.Errorf("Error %v creating connection", err)
+	}
+	lConn.SetTimeout(time.Duration(l.Config.ConnectionTimeout) * time.Second)
+
+	err = lConn.Bind(l.Config.ServiceAccountUsername, l.Config.ServiceAccountPassword)
+	if err != nil {
+		return nil, fmt.Errorf("Error %v in ldap bind", err)
+	}
+
+	return lConn, nil
+}
+
+// GenerateToken generates token
+func (l *LClient) GenerateToken(jsonInput map[string]string) (model.Token, error) {
+	log.Info("Now generating Ldap token")
+	searchConfig = l.InitializeSearchConfig()
```

Comment:
This does not seem to be the correct place to call l.InitializeSearchConfig(). This means that searchConfig object will be initialized only if GenerateToken method is called. This is true that in a normal scenario, during enabling Access Control, GenerateToken will be called before any other method here.

But what if rancher-auth-service is restarted (this happens everytime a config setting changes)? auth_server.reload() will initialize the ad_provider, but your ldap_client's searchConfig will be nil.

Any call to say GetIdentities after that will hit a panic due to nil searchConfig.

ldapClient should be initialized from the ad_provider when loadConfig() is called to set the auth config.